gmt-4.5.14/src/x2sys/x2sys_datalist.1
.TH X2SYS_DATALIST 1 "1 Nov 2015" "GMT 4.5.14" "Generic Mapping Tools"
.SH NAME
x2sys_datalist \- A generic data-extractor for ASCII or binary files
.SH SYNOPSIS
\fBx2sys_datalist\fP \fItrack(s)\fP \fB\-T\fP\fITAG\fP [ \fB\-A\fP ] [ \fB\-F\fP\fIname1,name2,...\fP ] 
[ \fB\-H\fP[\fBi\fP][\fInrec\fP] ] [ \fB\-L\fP[\fIcorrtable\fP] ] [ \fB\-R\fP\fIwest\fP/\fIeast\fP/\fIsouth\fP/\fInorth\fP[\fBr\fP] ]  [ \fB\-S\fP ] [ \fB\-V\fP ] [ \fB\-bo\fP[\fBs\fP|\fBS\fP|\fBd\fP|\fBD\fP[\fIncol\fP]|\fBc\fP[\fIvar1\fP\fB/\fP\fI...\fP]] ] [ \fB\-m\fP[\fIflag\fP] ]
.SH DESCRIPTION
\fBx2sys_datalist\fP reads one or more files and produces a single ASCII [or binary] table. 
The files can be of any format, which must be described and passed with the
\fB\-T\fP option.  You may limit the output to a geographic region, and insist
that the output from several files be separated by a multiple segment header.
Only the named data fields will be output [Default selects all columns].
.TP
\fItracks\fP
Can be one or more ASCII, native binary, or COARDS netCDF 1-D data files.
To supply the data files via a text file with a list of tracks (one per record),
specify the name of the track list after a leading equal-sign
(e.g., =tracks.lis).  If the names are missing their file extension we will
append the suffix specified for this \fITAG\fP.  Track files will be searched
for first in the current directory and second in all directories listed in
\fB$X2SYS_HOME\fP/\fITAG\fP/\fITAG\fP_paths.txt (if it exists). [If \fB$X2SYS_HOME\fP
is not set it will default to \fB$GMT_SHAREDIR\fP/x2sys]. (Note: MGD77 files will
also be looked for via \fBMGD77_HOME\fP/mgd77_paths.txt and *.gmt files will be
searched for via \fB$GMT_SHAREDIR\fP/mgg/gmtfile_paths).
.TP
\fB\-T\fP
Specify the x2sys \fITAG\fP which tracks the attributes of this data type.
.SH OPTIONS
No space between the option flag and the associated arguments.
.TP
\fB\-A\fP
Eliminate COEs by distributing the COE between the two tracks in proportion to track weight.
These (dist, adjustment) spline knots files for each track and data column
are called \fItrack.column\fP.adj and are expected to be in the \fB$X2SYS_HOME\fP/\fITAG\fP directory.
The adjustments are only applied if the corresponding adjust file can be found [No residual adjustments]
.TP
\fB\-F\fP
Give a comma-separated sub-set list of column names defined in the definition
file. [Default selects all data columns].
.TP
\fB\-H\fP
Input file(s) has header record(s).  If used, the default number of header records is \fBN_HEADER_RECS\fP.
Use \fB\-Hi\fP if only input data should have header records [Default will write out header records if the
input data have them]. Blank lines and lines starting with # are always skipped.
.TP
\fB\-L\fP
Apply optimal corrections to columns where such corrections are available.  Append the correction
table to use [Default uses the correction table \fITAG\fP_corrections.txt which is expected to reside
in the \fB$X2SYS_HOME\fP/\fITAG\fP directory].
For the format of this file, see CORRECTIONS below.
.TP
\fB\-R\fP
\fIwest, east, south,\fP and \fInorth\fP specify the Region of interest, and you may specify them
in decimal degrees or in [+-]dd:mm[:ss.xxx][W|E|S|N] format.  Append \fBr\fP if lower left and upper right
map coordinates are given instead of w/e/s/n.  The two shorthands \fB\-Rg\fP and \fB\-Rd\fP stand for global domain
(0/360 and -180/+180 in longitude respectively, with -90/+90 in latitude). Alternatively, specify the name
of an existing grid file and the \fB\-R\fP settings (and grid spacing, if applicable) are copied from the grid.
For Cartesian data just give \fIxmin/xmax/ymin/ymax\fP.  This option limits
the COEs to those that fall inside the specified domain.
.TP
\fB\-S\fP
Suppress output records where all the data columns are NaN [Default will output all records].
.TP
\fB\-V\fP
Selects verbose mode, which will send progress reports to stderr [Default runs "silently"].
.TP
\fB\-bo\fP
Selects binary output.
Append \fBs\fP for single precision [Default is \fBd\fP (double)].
Uppercase \fBS\fP or \fBD\fP will force byte-swapping.
Optionally, append \fIncol\fP, the number of desired columns in your binary output file.
.TP
\fB\-m\fP
Output a multisegment header between data from each track.  Note this option does not
imply anything about the input file; that information is conveyed via the system tag (\fB\-T\fP).
.SH EXAMPLES
To extract all data from the old-style MGG supplement file c2104.gmt, recognized by the tag GMT:
.br
.sp
\fBx2sys_datalist\fP c2104.gmt \fB\-T\fP GMT > myfile
.br
.sp
To make lon,lat, and depth input for \fBblockmean\fP and \fBsurface\fP using all the files
listed in the file tracks.lis and define by the tag TRK, but only the data that are inside the specified
area, and make output binary, run
.br
.sp
\fBx2sys_datalist\fP =tracks.lis \fB\-T\fP TRK \fB\-F\fP lon,lat,depth \fB\-R\fP-40/-30/25/35 \fB\-bo\fP > alltopo_bin.xyz
.SH CORRECTIONS
The correction table is an ASCII file with coefficients and parameters needed to carry out corrections.
This table is usually produced by \fBx2sys_solve\fP.
Comment records beginning with # are allowed.  All correction records are of the form
.br
.sp
\fItrackID observation correction\fP
.br
.sp
where \fItrackID\fP is the track name, \fIobservation\fP is one of the abbreviations for an observed field
contained in files under this TAG, and \fIcorrection\fP consists of one or more white-space-separated \fIterm\fPs that will be
\fBsubtracted\fP from the observation before output.  Each \fIterm\fP must have this exact syntax:
.br
.sp
\fIfactor\fP[*[\fIfunction\fP]([\fIscale\fP](\fIabbrev\fP[-\fIorigin\fP]))[^\fIpower\fP]]
.br
.sp
where terms in brackets are optional (the brackets themselves are not used but regular parentheses must
be used exactly as indicated).  No spaces are allowed except between \fIterm\fPs. The \fIfactor\fP is the amplitude
of the basis function, while the optional \fIfunction\fP can be one of sin, cos, or exp. The
optional \fIscale\fP and \fIorigin\fP can be used to translate the argument (before giving it to the optional
function).  The argument \fIabbrev\fP is one of the abbreviations for columns known to this TAG.
However, it can also be one of the three auxiliary terms \fBdist\fP (for along-track distances), \fBazim\fP for
along-track azimuths, and \fBvel\fP (for along-track speed); these are all sensitive to the \fB\-C\fP and \fB\-N\fP
settings used when defining the TAB; furthermore, \fBvel\fP requires \fBtime\fP to be present in the data.  If \fIorigin\fP is
given as \fBT\fP it means that we should replace it with the value of \fIabbrev\fP for the very
first record in the file (this is usually only done for \fItime\fP).  If the first data record entry is
NaN we revert \fIorigin\fP to zero.  Optionally, raise the entire expression to the given \fIpower\fP,
before multiplying by \fIfactor\fP.  The following is an example of fictitious corrections to the
track ABC, implying the \fBz\fP column should have a linear
trend removed, the field \fBobs\fP should be corrected by a strange dependency on latitude,
\fBweight\fP needs to have 1 added (hence correction is given as -1), and \fBfuel\fP should be
reduced by a linear distance term:
.br
.sp
ABC z	7.1	1e-4*((time-T))
.br
ABC obs	0.5*exp(-1e-3(lat))^1.5
.br
ABC weight	-1
.br
ABC fuel 0.02*((dist))
.SH "SEE ALSO"
.IR blockmean (1),
.IR GMT (1),
.IR surface (1),
.IR x2sys_init (1),
.IR x2sys_datalist (1),
.IR x2sys_get (1),
.IR x2sys_list (1),
.IR x2sys_put (1),
.IR x2sys_report (1),
.IR x2sys_solve (1)


gmt-4.5.14/ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 10331 2015-10-29 19:41:56Z pwessel $
-----------------------------------------------------------------------------------------

-----------------------------------------------------------------------------------------
2015-11-01	pwessel
	* GMT 4.5.14 finalized and released
-----------------------------------------------------------------------------------------
2015-10-29	pwessel
	* meca/psmeca.c:	Enforce rakes and strikes to be in -180/+180 range.
				Issue #799, for gmt4 only as not a problem in gmt5.

2015-07-28	pwessel
	* meca/psmeca.c:	Be able to read any value x and y coordinates

2015-06-24	pwessel
	* gmt_support.c:	Fix bug in jump near Dateline in gmt_inonout_sphpol_count,
				affecting inside/outside spherical cases, reported in message-2219

2015-03-18	pwessel
	* share/cpt/GMT_wysiwyg.cpt:	Fix typo in blue for high end [Issue #689].

2015-01-16	pwessel
	* gmt_plot.c:		Fix issue # 662 with graph arrows for negative scales.

2015-01-10	pwessel
	* gmt_plot.c:		Flipping the n/s sides array when lon and -y axes
				fixes issue #520

2015-01-01	pwessel
	* configure.ac:		Moved to 4.5.14 and back to svn release only.
-----------------------------------------------------------------------------------------
2015-01-01	pwessel
	* GMT 4.5.13 finalized and released [GSHHG 2.3.4 was released on Jan 1 as well]
-----------------------------------------------------------------------------------------
2014-12-28	pwessel
	* gmt_map.c:		Needed separate GMT_cartesian_overlap for non-geographic
				projections.  Fixes issue # 657.

2014-09-29	pwessel
	* gmt_plot.c:		Left a PSL variable undefined when grdimage plotted a grid
				with x = longitude and y = Cartesian.

2014-09-09	pwessel
	* grdfilter.c:		Fix y-shifts when -I sets smaller output.  Issue #616.
	* gmt_bcr.c:		Fix problem with points slightly outside causing an
				access violation; see issue #620.

2014-08-29	pwessel
	* sample1d.c:		Fix resampling for decreasing t values.  Issue #608.

2014-08-05	pwessel
	* gmt_map.c:		Could run into trouble clipping polygons for non-periodic
				maps.  Added more checks in GMT_wesn_clip.

2014-07-31	pwessel
	* gmt_map.c:		Rect clipping with -R<w/s/e/n>r could introduce stray lines
				when dataset has features about 180 degrees away from
				the user's area. See issue # 594.

2014-07-30	pwessel
	* configure.ac:		Set default GSHHG version to 2.3.2 (released August 1, 2014)
	* psxy[z].c:		Fixed wrap-around issue in fault lines.

2014-07-27	pwessel
	* gmtselect.c:		The -Z option needed to pass records with NaNs.

2014-07-19	pwessel
	* x2sys/x2sys_solve.c:	Fix problems that arise when a track has less crossovers
				than model parameters.
	* gmtmath.c:		Fix issue #123 with files and -C.

2014-07-01	pwessel
	* configure.ac:		Set default GSHHG version to 2.3.1 (released July 1, 2014)
	
2014-06-28	pwessel
	* mgd77/mgd77sniffer.c:	Needed to check that grid values used in decimate function
	 			also were within reasonable range.

2014-05-26	pwessel
	* gmt_map.c:		The map region clip path for -JE/e was hardwired to a full
				circle and did not care about -R setting.
				When -R...r and -Je/E was given the search for enclosing
				geographic boundaries failed when either the S or N pole
				was the projection's antipole.
2014-05-25	pwessel
	* grdmath.c:		Fix issue #566 bug in grdmath's KM2DEG operator.

2014-03-19	pwessel
	* gmtselect.c:		The -Fpolygon option might miss points because the io-machinery
				would sometimes set the min/max longitudes found incorrectly.

2014-03-14	pwessel
	* gmt_shore.c:		Give more error feedback if nc_open fails due to old library.

2014-03-01	pwessel
	* configure.ac:		Moved to 4.5.13 and back to svn release only.

-----------------------------------------------------------------------------------------
2014-03-01	pwessel
	* GMT 4.5.12 finalized and released [GSHHG 2.3.0 was released on Feb 1]
-----------------------------------------------------------------------------------------
2014-02-17	fwobbe
	* gmt_init.c etc.:	File locking -- if enabled via -DFLOCK -- is
				non-compulsory now.  This means if fcntl fails for some reason,
				GMT continues to access the file.  Hence, it is recommended to
				configure GMT with --enable-flock even if fcntl might be buggy
				on some ancient NFS shares.

2014-02-17	pwessel
	* gmt_map.c:		Size for memory allocation cast the product of nx * ny to
				size_t but overflow had already happen as calculation was
				done using ints, then cast to size_t. Must cast nx and ny
				separately.
2014-02-10	jluis
	* misc/gmt2kml.c:	Make sure the default style IDs are unique for each process.

2014-02-07	jluis
	* ps2raster.c:		Make sure that images in KML files do not carry a directory name

2014-02-03	pwessel
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-02-01	pwessel
	* configure.ac:		Since GSHHG 2.3.0 was released today, make that the default

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then
				a final extra character \0 would be added and exceed array length.
				Thanks to Joachim Saul for pointing it out.

2014-01-22	pwessel
	* pscoast.c:		The processing of -I -N levels and pens could get out of sync
				so that the wrong pen was used for the specified feature.

2014-01-16	remko
	* pslib.c:		Take superscripts and subscripts into account when determining
				dimensions of text box.

2014-01-10	pwessel
	* grdmath.c:		Had to let SDIST return km instead of degrees since it cannot
				do degrees for geodesics.  By doing km it is now aligned with
				what PDIST and LDIST returns for geographic data, and matches
				what GMT5 does.  Some scripts had to be modified to handle the
				new behavior.  If you use SDIST you may wish to check your usage.
				Also fixed bug in SDIST in that spherical selected geodesics

gmt-4.5.14/src/x2sys/x2sys_datalist.c
/*-----------------------------------------------------------------
 *	$Id: x2sys_datalist.c 10289 2014-12-28 21:17:06Z pwessel $
 *
 *      Copyright (c) 1999-2015 by P. Wessel
 *      See LICENSE.TXT file for copying and redistribution conditions.
 *
 *      This program is free software; you can redistribute it and/or modify
 *      it under the terms of the GNU General Public License as published by
 *      the Free Software Foundation; version 2 or any later version.
 *
 *      This program is distributed in the hope that it will be useful,
 *      but WITHOUT ANY WARRANTY; without even the implied warranty of
 *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *      GNU General Public License for more details.
 *
 *      Contact info: www.soest.hawaii.edu/pwessel
 *--------------------------------------------------------------------*/
/* x2sys_datalist will read one or several data files and dump their
 * contents to stdout in ascii or binary (double precision) mode.
 * Input data file formats are determined by the definition file
 * given by the -D option.
 *
 * Author:	Paul Wessel
 * Date:	15-JUN-2004
 * Version:	1.1, based on the spirit of the old xsystem code
 *
 */

#include "x2sys.h"

struct X2SYS_ADJUST {
	int n;
	double *d, *c;
};

GMT_LONG x2sys_load_adjustments (char *DIR, char *TAG, char *track, char *column, struct X2SYS_ADJUST **A);

int main (int argc, char **argv)
{
	char *fflags = CNULL, *TAG = CNULL, *correction_table = CNULL;
	char **trk_name = NULL, buffer[BUFSIZ];

	GMT_LONG i, j, k, bad, trk_no, n_data_col_out = 0;
	int n_tracks;

	GMT_LONG error = FALSE, suppress = FALSE, apply_corrections = FALSE, t_given, set_r = FALSE;
	GMT_LONG adjust = FALSE, cmdline_files, special_formatting = FALSE, xy_check, *adj_col = NULL;

	double **data = NULL, west, east, south, north, *out = NULL, correction = 0.0, aux_dvalue[N_GENERIC_AUX];
	double ds = 0.0, cumulative_dist, dist_scale = 1.0, dt, vel_scale = 1.0, adj_amount;

	struct X2SYS_INFO *s = NULL;
	struct X2SYS_FILE_INFO p;		/* File information */
	struct X2SYS_BIX B;
	struct MGD77_CORRTABLE **CORR = NULL;
	struct MGD77_AUX_INFO aux[N_MGD77_AUX];
	struct MGD77_AUXLIST auxlist[N_GENERIC_AUX] = {
		{ "dist",    MGD77_AUX_DS, 0, 0, "d(km)"},
		{ "azim",    MGD77_AUX_AZ, 0, 0, "azimuth"},
		{ "vel",     MGD77_AUX_SP, 0, 0, "v(m/s)"}
	};
	struct X2SYS_ADJUST **A = NULL;
	PFD GMT_azimuth_func;

	argc = (int)GMT_begin (argc, argv);
	for (i = strlen(argv[0]); i >= 0 && argv[0][i] != '/'; i--);
	X2SYS_program = &argv[0][i+1];	/* Name without full path */
	west = east = south = north = 0.0;
	t_given = FALSE;

	for (i = 1; i < argc; i++) {
		if (argv[i][0] == '-') {
			switch (argv[i][1]) {
				/* Common parameters */

				case 'H':
				case 'M':
				case 'R':
				case 'V':
				case 'b':
				case 'm':
				case '\0':
					error += GMT_parse_common_options (argv[i], &west, &east, &south, &north);
					break;

				/* Supplemental parameters */

				case 'A':
					adjust = TRUE;
					break;
				case 'F':
					fflags = &argv[i][2];
					break;
				case 'L':	/* Crossover correction table */
					correction_table = &argv[i][2];
					apply_corrections = TRUE;
					break;
				case 'S':
					suppress = TRUE;
					break;
				case 'T':
					TAG = &argv[i][2];
					t_given = TRUE;
					break;
				default:
					error = TRUE;
					break;
			}
		}
	}

	if (argc == 1 || error) {
		fprintf (stderr, "x2sys_datalist %s - listing of datafiles to stdout\n\n", X2SYS_VERSION);
		fprintf (stderr, "usage: x2sys_datalist <files> -T<TAG> [-A] [-F<fields>] [-L[<corrtable.txt>]]\n");
		fprintf (stderr, "\t[%s] [-S] [-V] [%s] [%s]\n\n", GMT_Rgeo_OPT, GMT_bo_OPT, GMT_mo_OPT);
		fprintf (stderr, "\t<files> is one or more datafiles, or give =<files.lis> for a file with a list of datafiles\n");
		fprintf (stderr, "\t-T <TAG> is the system tag for the data set.\n");
		fprintf (stderr, "\n\tOPTIONS:\n");
		fprintf (stderr, "\t-A Use any adjustment splines per track to redistribute COEs between tracks\n");
		fprintf (stderr, "\t   according to their relative weight [no adjustments].\n");
		fprintf (stderr, "\t-F is comma-separated list of column names to output [Default are all fields]\n");
		fprintf (stderr, "\t-L Subtract systematic corrections from the data. If no correction file is given,\n");
		fprintf (stderr, "\t   the default file <TAG>_corrections.txt in $X2SYS_HOME/<TAG> is assumed.\n");
		GMT_explain_option ('R');
		fprintf (stderr, "\t-S Suppress output records where all data columns are NaN [Output all records]\n");
		GMT_explain_option ('V');
		GMT_explain_option ('o');
		fprintf (stderr, "\t-m will write a multi-segment header between the output from each file\n");
		exit (EXIT_FAILURE);
	}

	if (!t_given) {
		fprintf (stderr, "%s: ERROR: Must specify -T\n", GMT_program);
		exit (EXIT_FAILURE);
	}

	if ((n_tracks = x2sys_get_tracknames (argc, argv, &trk_name, &cmdline_files)) == 0) {
		fprintf (stderr, "%s: No datafiles given!\n", GMT_program);
		exit (EXIT_FAILURE);		
	}

	if (!project_info.region_supplied) set_r = TRUE;

	x2sys_err_fail (x2sys_set_system (TAG, &s, &B, &GMT_io), TAG);

	if (fflags) x2sys_err_fail (x2sys_pick_fields (fflags, s), "-F");

	s->ascii_out = !GMT_io.binary[1];
	xy_check = (s->use_column[s->x_col[GMT_IN]] && s->use_column[s->y_col[GMT_IN]]);	/* TRUE means we requested lon,lat so we can check these for inside/outside */

	if (set_r) {
		west = B.x_min;		east = B.x_max;
		south = B.y_min;	north = B.y_max;
		if (!xy_check) {
			fprintf (stderr, "%s: Region specified but lon/lat (or x/y) columns not selected; -R ignored\n", GMT_program);
		}
	}
	if (project_info.region_supplied) {
		/* Supply dummy linear proj */
		project_info.projection = project_info.xyz_projection[0] = project_info.xyz_projection[1] = GMT_LINEAR;
		project_info.pars[0] = project_info.pars[1] = 1.0;
		if (s->geographic) project_info.degree[0] = project_info.degree[1] = TRUE;
		if (west < 0.0 && east < 0.0) {
			west += 360.0;
			east += 360.0;
		}
		GMT_err_fail (GMT_map_setup (west, east, south, north), "");
	}

	out = (double *) GMT_memory (VNULL, (size_t)s->n_fields, sizeof (double), "x2sys_datalist");

	for (i = 0; i < s->n_out_columns; i++) {	/* Set output formats */
		if (i == s->t_col[GMT_OUT])
			GMT_io.out_col_type[i] = GMT_IS_ABSTIME;
		else if (i == s->x_col[GMT_OUT])
			GMT_io.out_col_type[i] = (!strcmp (s->info[s->out_order[i]].name, "lon")) ? GMT_IS_LON : GMT_IS_FLOAT;
		else if (i == s->y_col[GMT_OUT])
			GMT_io.out_col_type[i] = (!strcmp (s->info[s->out_order[i]].name, "lat")) ? GMT_IS_LAT : GMT_IS_FLOAT;
		else
			GMT_io.out_col_type[i] = GMT_IS_FLOAT;

		if (s->info[s->out_order[i]].format[0] != '-') special_formatting = TRUE;
	}
	if (GMT_io.binary[GMT_OUT]) special_formatting = FALSE;

	if (suppress) {	/* Must count output data columns (except t, x, y) */
		for (i = n_data_col_out = 0; i < s->n_out_columns; i++) {
			if (i == s->t_col[GMT_OUT]) continue;
			if (i == s->x_col[GMT_OUT]) continue;
			if (i == s->y_col[GMT_OUT]) continue;
			n_data_col_out++;
		}
	}

	MGD77_Set_Unit (s->unit[X2SYS_DIST_SELECTION], &dist_scale, -1);	/* Gets scale which multiplies meters to chosen distance unit */
	MGD77_Set_Unit (s->unit[X2SYS_SPEED_SELECTION], &vel_scale,  -1);	/* Sets output scale for distances using in velocities */
	switch (s->unit[X2SYS_SPEED_SELECTION][0]) {
		case 'c':
			vel_scale = 1.0;
			break;
		case 'e':
			vel_scale /= dist_scale;			/* Must counteract any distance scaling to get meters. dt is in sec so we get  m/s */
			strcpy (auxlist[MGD77_AUX_SP].header, "v(m/s)");
			break;
		case 'k':
			vel_scale *= (3600.0 / dist_scale);		/* Must counteract any distance scaling to get km. dt is in sec so 3600 gives  km/hr */
			strcpy (auxlist[MGD77_AUX_SP].header, "v(km/hr)");
			break;
		case 'm':
			vel_scale *= (3600.0 / dist_scale);		/* Must counteract any distance scaling to get miles. dt is in sec so 3600 gives  miles/hr */
			strcpy (auxlist[MGD77_AUX_SP].header, "v(mi/hr)");
			break;
		case 'n':
			vel_scale *= (3600.0 / dist_scale);		/* Must counteract any distance scaling to get miles. dt is in sec so 3600 gives  miles/hr */
			strcpy (auxlist[MGD77_AUX_SP].header, "v(kts)");
			break;
	}
	switch (s->unit[X2SYS_DIST_SELECTION][0]) {
		case 'c':
			strcpy (auxlist[MGD77_AUX_SP].header, "d(user)");
			break;
		case 'e':
			strcpy (auxlist[MGD77_AUX_SP].header, "d(m)");
			break;
		case 'k':
			strcpy (auxlist[MGD77_AUX_SP].header, "d(km)");
			break;
		case 'm':
			strcpy (auxlist[MGD77_AUX_SP].header, "d(miles)");
			break;
		case 'n':
			strcpy (auxlist[MGD77_AUX_SP].header, "d(nm)");
			break;
	}

	switch (s->dist_flag) {
		case 0:	/* Cartesian */
			GMT_distance_func = GMT_cartesian_dist;
			GMT_azimuth_func  = GMT_az_backaz_cartesian;
			break;
		case 1:	/* Flat earth */
			GMT_distance_func = GMT_flatearth_dist_meter;
			GMT_azimuth_func  = GMT_az_backaz_flatearth;
			break;
		case 2:	/* Great circle */
			GMT_distance_func = GMT_great_circle_dist_meter;
			GMT_azimuth_func  = GMT_az_backaz_sphere;
			break;
		default:	/* Geodesic */
			GMT_distance_func = GMT_geodesic_dist_meter;
			GMT_azimuth_func  = GMT_az_backaz_geodesic;
			break;
	}
	
	if (apply_corrections) {	/* Load an ephemeral correction table */
		x2sys_get_corrtable (s, correction_table, n_tracks, trk_name, NULL, aux, auxlist, &CORR);
		if (auxlist[MGD77_AUX_SP].requested && s->t_col[GMT_IN] == -1) {
			fprintf (stderr, "%s: Selected correction table requires velocity which implies time (not selected)\n", GMT_program);
			MGD77_Free_Correction (CORR, n_tracks);
			x2sys_free_list (trk_name, n_tracks);
			exit (EXIT_FAILURE);
		}
	}

	if (adjust) {
		A = (struct X2SYS_ADJUST **) GMT_memory (VNULL, s->n_out_columns, sizeof (struct X2SYS_ADJUST *), GMT_program);
		adj_col = (GMT_LONG *) GMT_memory (VNULL, s->n_out_columns, sizeof (GMT_LONG), GMT_program);
	}
	
	for (trk_no = 0; trk_no < n_tracks; trk_no++) {

		if (gmtdefs.verbose) fprintf (stderr, "x2sys_datalist: Reading track %s\n", trk_name[trk_no]);

		x2sys_err_fail ((s->read_file) (trk_name[trk_no], &data, s, &p, &GMT_io, &k), trk_name[trk_no]);

		if (apply_corrections && s->t_col[GMT_IN] >= 0) MGD77_Init_Correction (CORR[trk_no], data);	/* Initialize origins if needed */

		if (adjust) {
			for (k = 0; k < s->n_out_columns; k++) adj_col[k] = x2sys_load_adjustments (X2SYS_HOME, TAG, trk_name[trk_no], s->info[s->out_order[k]].name, &A[k]);
		}

		if (GMT_io.multi_segments[GMT_OUT]) GMT_write_segmentheader (GMT_stdout, s->n_fields);

		cumulative_dist = 0.0;
		for (j = 0; j < p.n_rows; j++) {
			if (project_info.region_supplied && xy_check && GMT_map_outside (data[s->x_col[GMT_OUT]][j], data[s->y_col[GMT_OUT]][j])) continue;
			if (suppress) {
				for (k = bad = 0; k < s->n_out_columns; k++) {
					if (k == s->t_col[GMT_OUT]) continue;
					if (k == s->x_col[GMT_OUT]) continue;
					if (k == s->y_col[GMT_OUT]) continue;
					if (GMT_is_dnan (data[k][j])) bad++;
				}
				if (bad == n_data_col_out) continue;
			}
			if (auxlist[MGD77_AUX_AZ].requested) {
				if (j == 0)	/* Look forward at first point to get an azimuth */
					aux_dvalue[MGD77_AUX_AZ] = GMT_azimuth_func (data[s->x_col[GMT_OUT]][1], data[s->y_col[GMT_OUT]][1], data[s->x_col[GMT_OUT]][0], data[s->y_col[GMT_OUT]][0], FALSE);
				else		/* else go from previous to this point */
					aux_dvalue[MGD77_AUX_AZ] = GMT_azimuth_func (data[s->x_col[GMT_OUT]][j], data[s->y_col[GMT_OUT]][j], data[s->x_col[GMT_OUT]][j-1], data[s->y_col[GMT_OUT]][j-1], FALSE);
			}
			if (auxlist[MGD77_AUX_DS].requested) {
				ds = (j == 0) ? 0.0 : dist_scale * GMT_distance_func (data[s->x_col[GMT_OUT]][j], data[s->y_col[GMT_OUT]][j], data[s->x_col[GMT_OUT]][j-1], data[s->y_col[GMT_OUT]][j-1]);
				cumulative_dist += ds;
				aux_dvalue[MGD77_AUX_DS] = cumulative_dist;
			}
			if (auxlist[MGD77_AUX_SP].requested) {
				dt =  (j == 0) ? data[s->t_col[GMT_OUT]][1] - data[s->t_col[GMT_OUT]][0] : data[s->t_col[GMT_OUT]][j] - data[s->t_col[GMT_OUT]][j-1];
				aux_dvalue[MGD77_AUX_SP] = (GMT_is_dnan (dt) || dt == 0.0) ? GMT_d_NaN : vel_scale * ds / dt;
			}
			for (k = 0; k < s->n_out_columns; k++) {	/* Load output record */
				correction = (apply_corrections) ? MGD77_Correction (CORR[trk_no][k].term, data, aux_dvalue, j) : 0.0;
				if (adjust && adj_col[k]) {
					if (GMT_intpol (A[k]->d, A[k]->c, A[k]->n, 1, &aux_dvalue[MGD77_AUX_DS], &adj_amount, gmtdefs.interpolant)) {
						fprintf (stderr, "%s: Error interpolating adjustment for %s near row %ld - no adjustment made!\n", GMT_program, s->info[s->out_order[k]].name, j);
						adj_amount = 0.0;
					}
					correction -= adj_amount;
				}
				out[k] = data[k][j] - correction;	/* This loads out in the correct output order */
			}
			if (special_formatting)  {	/* use the specified formats */
				for (k = 0; k < s->n_out_columns; k++) {
					if (s->info[s->out_order[k]].format[0] == '-')
						GMT_ascii_output_one (GMT_stdout, out[k], k);
					else {
						if (!GMT_is_dnan (out[k])) {
							sprintf (buffer, s->info[s->out_order[k]].format, out[k]);
							GMT_fputs (buffer, GMT_stdout);
						}
						else
							GMT_fputs ("NaN", GMT_stdout);
					}
					(k == (s->n_out_columns - 1)) ? GMT_fputs ("\n", GMT_stdout) : GMT_fputs (gmtdefs.field_delimiter, GMT_stdout);
				}
			}
			else {
				GMT_output (GMT_stdout, s->n_out_columns, out);
			}
		}

		x2sys_free_data (data, s->n_out_columns, &p);
		for (k = 0; k < s->n_fields; k++) if (adjust && adj_col[k]) {
			GMT_free ((void *)A[k]->d);
			GMT_free ((void *)A[k]->c);
			GMT_free ((void *)A[k]);
		}
	}

	if (apply_corrections) MGD77_Free_Correction (CORR, n_tracks);

	x2sys_end (s);
	GMT_free ((void *)out);
	if (adjust) {
		GMT_free ((void *)A);
		GMT_free ((void *)adj_col);
	}
	x2sys_free_list (trk_name, n_tracks);
	
	GMT_end (argc, argv);

	exit (EXIT_SUCCESS);
}

GMT_LONG x2sys_load_adjustments (char *DIR, char *TAG, char *track, char *column, struct X2SYS_ADJUST **A)
{
	GMT_LONG n_fields, n_expected_fields = 2, n = 0, k, n_alloc = GMT_CHUNK;
	GMT_LONG type[2] = {GMT_IS_FLOAT, GMT_IS_FLOAT};
	double *in;
	char file[BUFSIZ];
	FILE *fp;
	struct X2SYS_ADJUST *adj;
	
	sprintf (file, "%s/%s/%s.%s.adj", DIR, TAG, track, column);
	if ((fp = GMT_fopen (file, "r")) == NULL) return FALSE;	/* Nuthin' to read */
	
	adj = (struct X2SYS_ADJUST *)GMT_memory (VNULL, 1, sizeof (struct X2SYS_ADJUST), GMT_program);
	adj->d = (double *)GMT_memory (VNULL, n_alloc, sizeof (double), GMT_program);
	adj->c = (double *)GMT_memory (VNULL, n_alloc, sizeof (double), GMT_program);
	for (k = 0; k < 2; k++) l_swap (type[k], GMT_io.in_col_type[k]);	/* Save original input type setting */
	while ((n_fields = GMT_input (fp, &n_expected_fields, &in)) >= 0 && !(GMT_io.status & GMT_IO_EOF)) {	/* Not yet EOF */
		adj->d[n] = in[0];
		adj->c[n] = in[1];
		n++;
		if (n == n_alloc) {
			n_alloc <<= 1;
			adj->d = (double *)GMT_memory ((void *)adj->d, n_alloc, sizeof (double), GMT_program);
			adj->c = (double *)GMT_memory ((void *)adj->c, n_alloc, sizeof (double), GMT_program);
		}
	}
	GMT_fclose (fp);
	adj->d = (double *)GMT_memory ((void *)adj->d, n, sizeof (double), GMT_program);
	adj->c = (double *)GMT_memory ((void *)adj->c, n, sizeof (double), GMT_program);
	adj->n = (int)n;
	*A = adj;
	for (k = 0; k < 2; k++) l_swap (GMT_io.in_col_type[k], type[k]);	/* Restore original input type setting */
	return (TRUE);
}
